input/tar2/tar_input.go

```diff
@@ -0,0 +1,163 @@
+	if spec.Type != Type {
+		panic(errors.ProgrammerError.New("This input implementation supports definitions of type %q, not %q", Type, spec.Type))
+	}
+	_, err := os.Stat(spec.URI)
+	if os.IsNotExist(err) {
+		panic(input.DataSourceUnavailableError.New("Input URI %q must be a tar file", spec.URI))
+	}
+	return &Input{
+		spec:          spec,
+		hasherFactory: sha512.New384,
+	}
+}
+
+func (i Input) Apply(destinationRoot string) <-chan error {
+	done := make(chan error)
+	go func() {
+		defer close(done)
+		try.Do(func() {
+			// do make a dir for untaring into.
+			// tars may specify permission and time bits for their top dir, but if not, we'll start with sane defaults.
+			err := os.MkdirAll(destinationRoot, 0755)
+			if err != nil {
+				panic(input.TargetFilesystemUnavailableIOError(err))
+			}
+			// daemon uid and gid are fine for now (they're always 0);
+			// forcing a constant time happens along with all the other dirs, since
+			// the bucket normalizes to include a root attribute set if the tar doesn't have one.
+
+			// open the tar file; preparing decompression as necessary
+			file, err := os.OpenFile(i.spec.URI, os.O_RDONLY, 0755)
+			if err != nil {
+				panic(input.DataSourceUnavailableIOError(err))
+			}
+			defer file.Close()
+			reader, err := Decompress(file)
+			if err != nil {
+				panic(input.DataSourceUnavailableError.New("could not start decompressing: %s", err))
+			}
+			tarReader := tar.NewReader(reader)
+
+			// unroll the tar, copying and accumulating data for integrity check
+			bucket := &fshash.MemoryBucket{}
+			walk(tarReader, destinationRoot, bucket, i.hasherFactory)
+
+			// hash whole tree
+			actualTreeHash, _ := fshash.Hash(bucket, i.hasherFactory)
+
+			// verify total integrity
+			expectedTreeHash, err := base64.URLEncoding.DecodeString(i.spec.Hash)
+			if !bytes.Equal(actualTreeHash, expectedTreeHash) {
+				done <- input.InputHashMismatchError.New("expected hash %q, got %q", i.spec.Hash, base64.URLEncoding.EncodeToString(actualTreeHash))
+			}
+		}).Catch(input.Error, func(err *errors.Error) {
+			done <- err
+		}).CatchAll(func(err error) {
+			// All errors we emit will be under `input.Error`'s type.
+			// Every time we hit this UnknownError path, we should consider it a bug until that error is categorized.
+			done <- input.UnknownError.Wrap(err).(*errors.Error)
+		}).Done()
+	}()
+	return done
+}
+
+func walk(tr *tar.Reader, destBasePath string, bucket fshash.Bucket, hasherFactory func() hash.Hash) error {
+	for {
+		thdr, err := tr.Next()
+		if err == io.EOF {
+			break // end of archive
+		}
+		if err != nil {
+			panic(input.DataSourceUnavailableError.New("corrupt tar: %s", err))
+		}
+		hdr := fs.Metadata(*thdr)
+		// filter/sanify values:
+		// names must be clean, relative dot-slash prefixed, and dirs slash-suffixed
+		// times should never be go's zero value; replace those with epoch
+		hdr.Name = path.Clean(hdr.Name)
+		if hdr.Name != "." {
+			hdr.Name = "./" + hdr.Name
+		}
```

Comment:
IMO, I would resolve this thread by adding a line to the comment block saying the path is NOT escape-safe at this point & that `fs.PlaceFile` handles escape attempts.

This will guard against us possibly adding code later that is not safe against escape attempts; I want to be really clear which parts of the system are handling safe or unsafe data.